### PR TITLE
[JUJU-4211] Remove support for podspec charms

### PR DESCRIPTION
**Pull request by wallyworld**
Operator style (podspec) charms are no longer supported in Juju4. This PR removes support for deploying them.
The relevant facades and workers are removed totally (eg caasoperatorprovisioner). Other facades like the caasunitprovisioner has had a haircut to remove methods which were only used by the old workers. And the k8s provider code has had a lot of now obsolete methods and tests removed. 
There were 2 firewallers - the legacy one is removed "sidecar" variant renamed to take its place.

## QA steps

bootstrap to microk8s
deploy snappass-test
scale-application snappass-test 2
deploy mysql-k8s

run status --storage to check that storage is created for mysql

### CODE REVIEW

File: apiserver/allfacades.go [hunk 1]
```diff
@@ -48,8 +47,11 @@ import (
 	"github.com/juju/juju/apiserver/facades/client/modelupgrader"
 	"github.com/juju/juju/apiserver/facades/client/pinger"
 	"github.com/juju/juju/apiserver/facades/client/secretbackends"
+	"github.com/juju/juju/apiserver/facades/controller/caasunitprovisioner"
 	"github.com/juju/juju/apiserver/facades/controller/crossmodelsecrets"
```
Comment (hpidcock) [inline apiserver/allfacades.go:52]: remove newline

File: api/controller/caasunitprovisioner/client.go [hunk 2]
```diff
@@ -9,36 +9,27 @@ import (
...
 	"github.com/juju/juju/rpc/params"
-	"github.com/juju/juju/storage"
 )
 
 // Client allows access to the CAAS unit provisioner API endpoint.
 type Client struct {
 	facade base.FacadeCaller
-	*charmscommon.CharmInfoClient
-	*charmscommon.ApplicationCharmInfoClient
+	//*charmscommon.CharmInfoClient
```
Comment (hpidcock) [inline api/controller/caasunitprovisioner/client.go:20]: should these just be removed?

File: api/controller/caasunitprovisioner/client.go [hunk 3]
```diff
@@ -9,36 +9,27 @@ import (
...
+	//*charmscommon.ApplicationCharmInfoClient
 }
 
 // NewClient returns a client used to access the CAAS unit provisioner API.
 func NewClient(caller base.APICaller) *Client {
 	facadeCaller := base.NewFacadeCaller(caller, "CAASUnitProvisioner")
-	charmInfoClient := charmscommon.NewCharmInfoClient(facadeCaller)
-	appCharmInfoClient := charmscommon.NewApplicationCharmInfoClient(facadeCaller)
+	//charmInfoClient := charmscommon.NewCharmInfoClient(facadeCaller)
```
Comment (hpidcock) [inline api/controller/caasunitprovisioner/client.go:27]: should these just be removed?

**Review by hpidcock — APPROVED** [review]
What a read... 🔥.

Thanks for resecting all this.

File: api/controller/caasunitprovisioner/client.go (hunk 2 above)
Comment (wallyworld) [reply api/controller/caasunitprovisioner/client.go:20]: yup, meant to do that

**Comment by jnsgruk** [comment]
@wallyworld - we should probably also remove the "application-config" within `config` while we're doing this. The only question that remains is what to do with `trust`, as that's the only attribute in there that is not just related to pod spec IIRC.

@manadart WDYT?

**Comment by wallyworld** [comment]
> @wallyworld - we should probably also remove the "application-config" within `config` while we're doing this. The only question that remains is what to do with `trust`, as that's the only attribute in there that is not just related to pod spec IIRC.
> 
> @jameinel @manadart WDYT?

We do have plans to model trust better, eg we've discussed providing charms with a vocab to allow them to declare what they want and trust would grant that.
The other application config type attributes are for things like service type, external ips, various load balancer attributes etc - attributes of the k8s service resource which is tied to (and named after) the application. These weren't brought across to the sidecar world, but we need to decide how we want to support this type of configuration moving forward. I've kept those configs in this PR on purpose so we can discuss before removing if that's what is decided.

**Comment by jnsgruk** [comment]
Yeh I agree that trust should be somewhere else.

My preference would be to eliminate "application-config" completely - even if we do eventually support the k8s service stuff, it won't be like this I don't think. Seems like a good opportunity to clean this slightly messy modelling up 🙂

**Comment by wallyworld** [comment]
> Yeh I agree that trust should be somewhere else.
> 
> My preference would be to eliminate "application-config" completely - even if we do eventually support the k8s service stuff, it won't be like this I don't think. Seems like a good opportunity to clean this slightly messy modelling up slightly_smiling_face

k8s app config is gone. We just have trust till we get to reimplement it.

**Pull request merged**